src/crawl.js

```diff
@@ -0,0 +1,107 @@
+/*
+eslint no-await-in-loop: "off",
+no-constant-condition: ["error", { "checkLoops": false }]
+*/
+
+import jsonld from 'jsonld';
+import assert from 'assert';
+import deepEqual from 'deep-equal';
+import { deepClone, assertValidNode } from './utils/common';
+import { queryNextLookup, dereferenceFromIPFS, parseRDFDocument } from './utils/rdf';
+import { inferh } from './utils/cd';
+import { canon } from './utils/canonicalize';
+import { Namer, fromJsonldjsCg } from './utils/claimgraph';
+
+// Crawl the rdf dataset composed of DIDs and turtle documents on ipfs. Return the graph
+// representing all knowlege obtained while crawling.
+export async function crawl(
+  initialFacts,
+  rules,
+  curiosityQuery,
+  resolveGraph,
+) {
+  // namer is used to ensure blank node hygene whenever adding new claimgraphs
+  const namer = new Namer();
+
+  let facts = deepClone(initialFacts);
+  namer.reallocateNames(facts);
+
+  const lookedup = new Set();
+  const marknew = (term) => {
+    // Add term to lookedup.
+    // Return whether the term was already present in the set.
+    // like https://doc.rust-lang.org/std/collections/btree_set/struct.BTreeSet.html#method.insert
+    const str = canon(term);
+    const isnew = !lookedup.has(str);
+    lookedup.add(str);
+    return isnew;
+  };
+
+  while (true) {
+    // reason over what we have so far
+    facts = facts.concat(inferh(facts, rules));
+
+    // lookup any interesting documents
+    const interesting = await queryNextLookup(facts, curiosityQuery);
+    const novel = interesting.filter(marknew);
+    const newfacts = [...await Promise.all(novel.map(resolveGraph))];
```

Comment:
why use spread syntax to create a new array? existing array returned by `await Promise.all(novel.map(resolveGraph))` should suffice

Reply:
Fixed 915cbbcd40c4195b3c092ad42f6861b1fdaa2b48

Reply:
I just remembered why I did that. Its because I need to check the length of `newfacts`. [Promse.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) returns an `iterator` but that iterator is not guaranteed to be an array.

Reply:
I could loop over the iterable to count it's elements but I think I should just change it back. I'm optimizing for readability here.

Reply:
Reverted 915cbbc in f73c833 . Thanks for pointing it out though.